### ISSUE TITLE
curdoc.clear() doesn't work anymore on 3.x

### ISSUE THREAD

**Issue by johann006** [bokeh/bokeh#12628]
### Software versions

Python version      :  3.9.10 (v3.9.10:f2f3f53782, Jan 13 2022, 16:55:45) 
IPython version     :  (not installed)
Tornado version     :  6.2
Bokeh version       :  2.4.3
BokehJS static path :  /Users/johannpayen/Dropbox/Dev Python/Dev Python NextOne/venv/lib/python3.9/site-packages/bokeh/server/static
node.js version     :  (not installed)
npm version         :  (not installed)
Operating system    :  macOS-12.6-arm64-arm-64bit

### Browser name and version

chrome

### Expected behavior

curdoc().clear() clear the document in Bokeh 2.x, since 3.x it stopped working.

There are other issues related : 
curdoc().title works only once
curdoc().add_root() works only once also

### Observed behavior

nothing seems to happen from the user point of view.

### Example code

```Python
import numpy as np

from bokeh.io import curdoc
from bokeh.layouts import column, row
from bokeh.models import ColumnDataSource, Slider, TextInput
from bokeh.plotting import figure

# Set up data
N = 200
x = np.linspace(0, 4*np.pi, N)
y = np.sin(x)
source = ColumnDataSource(data=dict(x=x, y=y))


# Set up plot
plot = figure(height=400, width=400, title="my sine wave",
              tools="crosshair,pan,reset,save,wheel_zoom",
              x_range=[0, 4*np.pi], y_range=[-2.5, 2.5])

plot.line('x', 'y', source=source, line_width=3, line_alpha=0.6)


# Set up widgets
text = TextInput(title="title", value='my sine wave')
offset = Slider(title="offset", value=0.0, start=-5.0, end=5.0, step=0.1)
amplitude = Slider(title="amplitude", value=1.0, start=-5.0, end=5.0, step=0.1)
phase = Slider(title="phase", value=0.0, start=0.0, end=2*np.pi)
freq = Slider(title="frequency", value=1.0, start=0.1, end=5.1, step=0.1)


# Set up callbacks
def update_title(attrname, old, new):
    plot.title.text = text.value
    if text.value =="clear":
        curdoc().clear()

text.on_change('value', update_title)

def update_data(attrname, old, new):

    # Get the current slider values
    a = amplitude.value
    b = offset.value
    w = phase.value
    k = freq.value

    # Generate the new curve
    x = np.linspace(0, 4*np.pi, N)
    y = a*np.sin(k*x + w) + b

    source.data = dict(x=x, y=y)

for w in [offset, amplitude, phase, freq]:
    w.on_change('value', update_data)


# Set up layouts and add to document
inputs = column(text, offset, amplitude, phase, freq)

curdoc().add_root(row(inputs, plot, width=800))
curdoc().title = "Sliders"
```

**Comment by bryevdv** [comment]
Not seeing any JS console errors when the `clear` is issued, just nothing happens. More investigation will be needed.

**Comment by AdriaanRol** [comment]
Has there been any further investigation on this? 

As a new bokeh user with version 3.0.3 I can confirm that this is not working for me either. 
---
EDIT
I have looked a bit deeper into this, and it appears that after calling `curdoc().clear()`, the document becomes unresponsive. That is, all the elements remain visible, but the actions associated to the widgets (e.g., on-click,) stop working. 

Interestingly enough, a periodic callback (does keep working). 

Below a minimal working example. 

```python

from bokeh.layouts import column
from bokeh.models.widgets import Button, Select
from bokeh.io import curdoc
from datetime import datetime


def add_select():
    print("add")
    feature = Select(value="feat", options=["a"])
    dynamic_col.children.insert(-1, feature)


def add_clear():
    feature = Select(value="feat", options=["a"])
    print("clearing")  # print added to show

    # curdoc().clear() is broken in 3.0.x
    curdoc().clear()


# the add on select stops working after clear has been called
# but the buttons remain visible.
b1 = Button(label="Add condition", button_type="success")
b1.on_click(add_select)

b2 = Button(label="Clear", button_type="success")
b2.on_click(add_clear)


def print_test_statement():
    # added for debugging, shows that the periodic callback keeps working after
    # clear is called
    print(datetime.now())


curdoc().add_periodic_callback(print_test_statement, period_milliseconds=1000)

dynamic_col = column(b1, b2)
curdoc().add_root(dynamic_col)
```

**Comment by bryevdv** [comment]
> Has there been any further investigation on this?

@AdriaanRol There is never any reason or need to ask this on an OSS tracker. If there were more information to report, it would already be here. 

Just to set expectations, given that:

* usage of `curdoc().clear()` does not actually seem to be very commonplace
* a workaround exists&mdash;create a "top level" layout like `column` and clear that out instead

there are probably quite a few other issues that will be considered higher priority than this in the short term.

**Comment by johann006** [comment]
Hello,

 Just to inform you that it is not only clear() that is not working anymore, but I’ve found that also add_next_tick_callback, add_root (if you you to add something dynamically)…
And sorry but I didn’t find any workaround for all these functions that I use.
I understand that this is not high priority, and I can stay on bokeh 2.x for my apps to continue working.But if you don’t plan to fix them, please let us know, as I’ll have to change all the design of my applications, which means weeks or months of work.

Best Regards

**Comment by bryevdv** [comment]
@johann006 I'm not sure what to tell you. We're not "not-planning" to fix anything, but Bokeh is a volunteer open source project, with many more issues than contributors. I can't offer you any specific guarantee about when someone might choose to spend their time on this issue, because I am not anyone's boss. As is always the case with OSS, the best way to accelerate the timeline of any given issue is to decide to become involved. 

Regarding:

>  Just to inform you that it is not only clear() that is not working anymore, but I’ve found that also add_next_tick_callback, add_root (if you you to add something dynamically)…
And sorry but I didn’t find any workaround for all these functions that I use.

There are only issues when `curdoc().clear()` is called, is that not correct? [1] I am suggesting (as a WAR): don't call `curdoc().clear()` at all, so then there are no problems with those other things. Instead:
```python
page = column() # permanent top-level layout container
 
curdoc().add_root(page)
```
Then add and delete children of `page`, as an alternative to calling `curdoc().clear()`. 

[1] And if that is *not correct*, i.e. there are issues with those callbacks even *without* a call to `curdoc().clear()` then that is not really apparent from the information above, and should really get a new separate issue.

**Comment by AdriaanRol** [comment]
Hi @bryevdv , thanks for getting back. Knowing about the workaround is quite valuable, thanks 👍 . 

I have some practical questions on the internals of how this works. I tried looking a bit into the sourcecode of the document class (and trying variations of what is behind `.clear()`), but I couldn't really figure out why it isn't working. 

The main thing that I'd like to understand is what the proper way of "deleting" things. I tried to manually call the commands in `.clear`, but it appears these had the same side-effects as the broken `.clear()` method. 

Understanding how to properly clear the list of children manually (even if it is only in a column or similar object) might help me understand the internals enough to make a contribution.

**Comment by bryevdv** [comment]
This seems to have been fixed at some point. Both of the codes above work as expected and clear the document now with 3.6.2
